core/ai.go

```diff
@@ -127,3 +132,102 @@ func ParseStepsFromModelID(modelID *string, defaultSteps float64) float64 {
+	}
+
+	n.Capabilities.mutex.Lock()
+	defer n.Capabilities.mutex.Unlock()
+	for aiCapability, aiConstraint := range aiConstraints {
+		_, capExists := n.Capabilities.constraints.perCapability[aiCapability]
+		if !capExists {
+			n.Capabilities.constraints.perCapability[aiCapability] = &CapabilityConstraints{
+				Models: make(ModelConstraints),
+			}
+		}
+
+		for modelId, modelConstraint := range aiConstraint.Models {
+			_, modelExists := n.Capabilities.constraints.perCapability[aiCapability].Models[modelId]
+			if modelExists {
+				n.Capabilities.constraints.perCapability[aiCapability].Models[modelId].Capacity += modelConstraint.Capacity
+			} else {
+				n.Capabilities.constraints.perCapability[aiCapability].Models[modelId] = &ModelConstraint{Warm: modelConstraint.Warm, Capacity: modelConstraint.Capacity}
+			}
+		}
+	}
+}
+
+// RemoveAICapabilities removes AI capabilities from the node.
+func (n *LivepeerNode) RemoveAICapabilities(caps *Capabilities) {
+	aiConstraints := caps.PerCapability()
+	if aiConstraints == nil {
+		return
+	}
+
+	n.Capabilities.mutex.Lock()
+	defer n.Capabilities.mutex.Unlock()
+	for capability, constraint := range aiConstraints {
+		_, ok := n.Capabilities.constraints.perCapability[capability]
+		if ok {
+			for modelId, modelConstraint := range constraint.Models {
+				_, modelExists := n.Capabilities.constraints.perCapability[capability].Models[modelId]
+				if modelExists {
+					n.Capabilities.constraints.perCapability[capability].Models[modelId].Capacity -= modelConstraint.Capacity
+					if n.Capabilities.constraints.perCapability[capability].Models[modelId].Capacity <= 0 {
+						delete(n.Capabilities.constraints.perCapability[capability].Models, modelId)
+					}
+				} else {
+					glog.Errorf("failed to remove AI capability capacity, model does not exist pipeline=%v modelID=%v", capability, modelId)
+				}
+			}
+		}
+	}
+}
+
+func (n *LivepeerNode) ReserveAICapability(pipeline string, modelID string) error {
+	cap, err := PipelineToCapability(pipeline)
+	if err != nil {
+		return err
+	}
+
+	_, hasCap := n.Capabilities.constraints.perCapability[cap]
+	if hasCap {
+		_, hasModel := n.Capabilities.constraints.perCapability[cap].Models[modelID]
+		if hasModel {
+			n.Capabilities.mutex.Lock()
+			defer n.Capabilities.mutex.Unlock()
+			if n.Capabilities.constraints.perCapability[cap].Models[modelID].Capacity > 0 {
+				n.Capabilities.constraints.perCapability[cap].Models[modelID].Capacity -= 1
+			} else {
+				return fmt.Errorf("failed to reserve AI capability capacity, model capacity is 0 pipeline=%v modelID=%v", pipeline, modelID)
+			}
+			return nil
+		}
+		return fmt.Errorf("failed to reserve AI capability capacity, model does not exist pipeline=%v modelID=%v", pipeline, modelID)
+	}
+	return fmt.Errorf("failed to reserve AI capability capacity, pipeline does not exist pipeline=%v modelID=%v", pipeline, modelID)
+}
+
+func (n *LivepeerNode) ReleaseAICapability(pipeline string, modelID string) error {
```

Comment:
This function should not exist

Reply:
Not following, please explain. Intent here was to start managing capacity for each pipeline/modelID and also work with workers having multiple GPUs serving the same pipeline/modelID behind one ai worker.